### PR TITLE
Use Ref and Vector in a few places in lapack and arpack

### PR DESCRIPTION
**Pull request by musm**
~~edit: looks like I introduced some bugs here in the lapack wrapper~~

Instead of trying to change everything, I limited the scope here. Taking the piecemeal approach (which will probably also be a lot easier to review too)

### CODE REVIEW

File: base/linalg/arpack.jl [hunk 1]
```diff
@@ -12,20 +12,20 @@ function aupd_wrapper(T, matvecA!::Function, matvecB::Function, solveSI::Functio
...
-    workd  = Array{T}(3*n)
-    workl  = Array{T}(lworkl)
-    rwork  = cmplx ? Array{TR}(ncv) : Array{TR}(0)
+    v     = Matrix{T}(n, ncv)
+    workd = Vector{T}(3*n)
+    workl = Vector{T}(lworkl)
+    rwork = cmplx ? Vector{TR}(ncv) : Vector{TR}(0)
 
     if isempty(v0)
-        resid  = Array{T}(n)
```
Comment (musm) [inline base/linalg/arpack.jl]: I don't quite understand the logic here with `info` in this function, why does it need to be initialized to zero or one.

Also some of the signatures in arpack look like they can be tightened up to `::Vector`

File: base/linalg/lapack.jl [hunk 2]
```diff
@@ -295,8 +295,8 @@ for (gebrd, gelqf, geqlf, geqrf, geqp3, geqrt, geqrt3, gerqf, getrf, elty, relty
                      work, &lwork, info)
                 chklapackerror(info[])
                 if i == 1
-                    lwork = BlasInt(real(work[1]))
-                    work = Array{$elty}(lwork)
+                    lwork = BlasInt(real(work[]))
+                    work = Ref{$elty}(lwork)
```
Comment (JeffBezanson) [inline base/linalg/lapack.jl]: I think you meant `Vector` here.

File: base/linalg/arpack.jl [hunk 3]
```diff
@@ -12,20 +12,20 @@ function aupd_wrapper(T, matvecA!::Function, matvecB::Function, solveSI::Functio
                       tol::Real, maxiter::Integer, mode::Integer, v0::Vector)
     lworkl = cmplx ? ncv * (3*ncv + 5) : (sym ? ncv * (ncv + 8) :  ncv * (3*ncv + 6) )
     TR = cmplx ? T.types[1] : T
-    TOL = Array{TR}(1)
+    TOL = Vector{TR}(1)
```
Comment (andreasnoack) [inline base/linalg/arpack.jl]: Not directly the topic of this PR, but I'm pretty sure that you can just adjust the driver functions and pass `tol` to them without wrapping. It is only an input argument so so `&` should be fine.

**Review by andreasnoack — APPROVED** [review]

**Comment by musm** [comment]
OSX is erroring with     Expression: ex.code == Base.UV_EMFILE
which I think is a known issue
https://github.com/JuliaLang/julia/issues/17626

**Comment by musm** [comment]
g2g

File: base/linalg/arpack.jl [hunk 4]
```diff
@@ -228,7 +228,7 @@ for (T, saupd_name, seupd_name, naupd_name, neupd_name) in
                   sizeof(howmny), sizeof(bmat), sizeof(evtype))
         end
 
-        function saupd(ido, bmat, n, which, nev, TOL::Array{$T}, resid::Array{$T}, ncv, v::Array{$T}, ldv,
+        function saupd(ido, bmat, n, which, nev, TOL::Ref{$T}, resid::Array{$T}, ncv, v::Array{$T}, ldv,
```
Comment (tkelman) [inline base/linalg/arpack.jl]: I think we'll need to revert the changes that were visible in the signature here, this broke MatrixNetworks.jl https://github.com/JuliaCI/pkg.julialang.org/blob/c868b6c9a62315ab23df2331ffd159c23cb29f2e/logs/MatrixNetworks_0.6.log and we're in feature freeze

Comment (musm) [inline base/linalg/arpack.jl]: How about changing them to `Union{Array{$T},Ref{$T}}`

Comment (tkelman) [inline base/linalg/arpack.jl]: might do the job, if it's good enough that `Pkg.test("MatrixNetworks")` passes with it then works for me

Comment (musm) [inline base/linalg/arpack.jl]: can I instead just submit a PR to matrix networks ?

Comment (andreasnoack) [inline base/linalg/arpack.jl]: As mentioned earlier, I don't think `TOL` needs to be a `Ref`. It is only [an input parameter](http://www.caam.rice.edu/software/ARPACK/UG/node136.html#SECTION001210000000000000000), so I think we should revert to the `Array{$T}` for the 0.6 release and then change it to `T` when 0.6 has branched.

Comment (StefanKarpinski) [inline base/linalg/arpack.jl]: How about allowing the future form with just `T` now in the 0.6 cycle? That won't break anyone's code and then people can write calls already the way it will be in the future.

Comment (musm) [inline base/linalg/arpack.jl]: > How about allowing the future form with just T now in the 0.6 cycle? That won't break anyone's code and then people can write calls already the way it will be in the future.

I'm a little confused here what would the signature of look like in that case? If I understand correctly the proposal is to change to 
function saupd(.....,`TOL::T`,....)
but how would that not break code that passes  `TOL::Array{T}`

Comment (StefanKarpinski) [inline base/linalg/arpack.jl]: Multiple dispatch? There can be multiple signatures for the same function...

Comment (musm) [inline base/linalg/arpack.jl]: I guess, although it seems unnecessarily complex. I think the first suggestion to just revert and then refactor later after 0.6 might be the best course of action.
